firedrake/preconditioners/asm.py

```diff
@@ -66,6 +66,89 @@ def applyTranspose(self, pc, x, y):
         self.asmpc.applyTranspose(x, y)
 
 
+class ASMStarPC(ASMPatchPC):
+    '''Patch-based PC using Star of mesh entities implmented as an
+    :class:`ASMPatchPC`.
+
+    ASMStarPC is an additive Schwarz preconditioner where each patch
+    consists of all DoFs on the topological star of the mesh entity
+    specified by `pc_star_construct_dim`.
+    '''
+
+    _prefix = "pc_star_"
+
+    # Want to override some PETSc default options, so duplicate this
+    def initialize(self, pc):
+        # Get context from pc
+        _, P = pc.getOperators()
+        dm = pc.getDM()
+        self.prefix = pc.getOptionsPrefix() + self._prefix
+
+        # Extract function space and mesh to obtain plex and indexing functions
+        V = get_function_space(dm)
+
+        # Obtain patches from user defined funtion
+        ises = self.get_patches(V)
+
+        # Create new PC object as ASM type and set index sets for patches
+        asmpc = PETSc.PC().create(comm=pc.comm)
+        asmpc.incrementTabLevel(1, parent=pc)
+        asmpc.setOptionsPrefix(self.prefix + "sub_")
+        asmpc.setOperators(*pc.getOperators())
+        asmpc.setType(asmpc.Type.ASM)
+        asmpc.setASMLocalSubdomains(len(ises), ises)
+
+        # Set default solver parameters
+        opts = PETSc.Options(asmpc.getOptionsPrefix())
```

Comment:
Shouldn't these be the default for all `ASMXPC` implementations? I propose moving these to the base class and not subclassing `initialise` in `ASMStarPC`.

Reply:
I'm not sure how relevant these are to `ASMLinesmoothPC`.  If the patch submatrix is diagonal or tridiagonal, then ilu == lu, and you wouldn't see the difference.  But for stars, you really notice that ILU(0) is a bad thing to do in general.

Likewise, setting `pc_asm_type` to `basic` doesn't have much of an effect in many cases, but is important sometimes (e.g., I see very bad results without it for Nedelec elements in parallel in 3D).

TLDR: probably we should check if that would change anything for `ASMLinesmoothPC`.  @colinjcotter ?

Reply:
Alternately, if they shouldn't be default, prefer to not replicate the full `initialize` code, but instead advertise a `postSetUp` function or something that subclasses can override.

Also, rather than poking the options database, it's better to do this programmatically.

After doing `asmpc.setType(ASM)` you can do:

```python
asmpc.setASMType(PETSc.PC.ASMType.BASIC)
ksp = asmpc.getASMSubKSP()
ksp.pc.setType(PETSc.PC.Type.LU)
ksp.pc.setFactorShift(shift_type=PETSc.Mat.FactorShiftType.NONZERO)
```

Reply:
Okay.  I'll wait to consider any effects on `ASMLinesmoothPC` before pushing changes, but will change to the programmatic style either in the base class, or as a `postSetUp` function.

Reply:
Let's bikeshed the name a bit. Maybe `set_defaults(self, asmpc)` ?

Reply:
Is `set_defaults` too generic?  I guess that the context is easily inferred by the class structure, but `set_asm_defaults` seems rather less cryptic to me...

Reply:
am happy with `set_asm_defaults` too.

Reply:
re: defaults. I don't know how PETSc deals with this, but is each KSP local to a process domain usually? Just thinking that the default LU is not parallel.

Reply:
The KSPs are all local.

Reply:
The latest push removes the duplication, and just puts all the options stuff back in `ASMPatchPC`.  I _think_ that's reasonable from this conversation, but am open to adding `set_asm_defaults` if that's preferred.

I was foiled, however, in trying to do this programmatically.  I've left Lawrence's code suggestion stubbed in here in the comments, but I get a catastrophic (and uninformative) fail when I uncomment Line 53 of `asm.py`, and I haven't been successful in uncovering why.  Any ideas?